XSConsoleData.py

```diff
@@ -528,17 +545,95 @@ def SaveToNTPConf(self):
         Auth.Inst().AssertAuthenticated()
 
-        file = None
         try:
-            file = open("/etc/chrony.conf", "w")
-            for other in self.ntp.othercontents([]):
-                file.write(other+"\n")
-            for server in self.ntp.servers([]):
-                file.write("server "+server+" iburst\n")
+            with open("/etc/chrony.conf", "w") as confFile:
+              for other in self.ntp.othercontents([]):
+                  confFile.write(other + "\n")
+              for server in self.ntp.servers([]):
+                  confFile.write("server " + server + " iburst\n")
         finally:
-            if file is not None: file.close()
             self.UpdateFromNTPConf()
 
+        # Force chronyd to update the time
+        if self.data['ntp']['method'] != "Disabled":
+            servers = self.data['ntp']['servers']
+            if self.data['ntp']['method'] == "DHCP":
+                servers = self.GetDHClientInterfaces()
+
+            if servers:
+                self.StopService("chronyd")
+                getoutput("chronyd -q 'server %s iburst'" % servers[0])  # Update the clock and exit
+                getoutput("hwclock -w")  # Sync hwclock with date set by chronyd
+                self.StartService("chronyd")
+
+    def AddDHCPNTP(self):
+        # Double-check authentication
+        Auth.Inst().AssertAuthenticated()
+
+        oldPermissions = os.stat("/etc/dhcp/dhclient.d/chrony.sh").st_mode
+        newPermissions = oldPermissions | (stat.S_IXUSR | stat.S_IXGRP | stat.S_IXOTH)
+        os.chmod("/etc/dhcp/dhclient.d/chrony.sh", newPermissions)
+
+        interfaces = self.GetDHClientInterfaces()
+        for interface in interfaces:
+            ntpServer = self.GetDHCPNTPServer(interface)
+
+            with open("/var/lib/dhclient/chrony.servers.%s" % interface, "w") as chronyFile:
+                chronyFile.write("%s iburst prefer\n" % ntpServer)
+
+    def ResetDefaultNTPServers(self):
+        # Double-check authentication
+        Auth.Inst().AssertAuthenticated()
+
+        Data.Inst().NTPServersSet(["0.centos.pool.ntp.org",
+                                    "1.centos.pool.ntp.org",
+                                    "2.centos.pool.ntp.org",
+                                    "3.centos.pool.ntp.org"])
+
+    def GetDHClientInterfaces(self):
+        (status, output) = getstatusoutput("ls /var/lib/xcp/ | grep leases")
+        if status != 0:
+          return []
+
+        dhclientFiles = output.splitlines()
+        pattern = "dhclient-(.*).leases"
+        interfaces = []
+        for dhclientFile in dhclientFiles:
+            match = re.match(pattern, dhclientFile)
+            if match:
+                interfaces.append(match.group(1))
+
+        return interfaces
+
+    def GetDHCPNTPServer(self, interface):
+        ntpServer = getoutput("grep ntp-servers /var/lib/xcp/dhclient-%s.leases | tail -1 | awk '{{ print ( $3 ) }}'" % interface).strip()[:-1]
```

Comment:
Likewise, this could be done in a more Pythonic way.

I agree that having this as a one-line shell script is nice, though.

As `awk` is also used by a number of other XS scripts, I see no hard objections, though.

PS: darker would format it in this way, reducing the column count from 143 to 99
```suggestion
        ntpServer = getoutput(
            "grep ntp-servers /var/lib/xcp/dhclient-%s.leases | tail -1 | awk '{{ print ( $3 ) }}'"
            % interface
        ).strip()[:-1]
```
Maybe that's better, don't know.